### PR TITLE
Chronos: add `quantize` method for TSPipeline 

### PR DESCRIPTION
**Pull request by liangs6212**

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 1]
```diff
@@ -338,6 +353,107 @@ def load(file_path):
...
+               | 1. data creator:
+               | a function that takes a config dictionary as parameter and
+               | returns a PyTorch DataLoader.
+               |
+               | 2. a bigdl.chronos.data.TSDataset:
+               | the TSDataset should follow the same operations as the training
+               | TSDataset used in `AutoTSEstimator.fit`.
+               |
+               | 3. A torch.utils.data.dataloader.DataLoader object for calibration.
+               | Required for static quantization.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:381]: Users should set the configs correctly (e.g. past_seq_len, ...). They can be found in `TSPipeline._best_config`

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 2]
```diff
@@ -338,6 +353,107 @@ def load(file_path):
...
+        :param timeout: Tuning timeout (seconds). Default to 0, which means early stop.
+               Combine with max_trials field to decide when to exit.
+        :param max_trials: Max tune times. Default to 1. Combine with timeout field to
+               decide when to exit. "timeout=0, max_trials=1" means it will try quantization
+               only once and return satisfying best model.
+        """
+        from torch.utils.data import DataLoader, TensorDataset
+        # check model support for quantization
+        from bigdl.chronos.autots.utils import check_quantize_available
+        check_quantize_available(self._best_model.model)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:406]: use a blacklist for the quantization availability.

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py (hunk 1 above)
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/autots/tspipeline.py:381]: Required for static quantization. can be moved ahead.

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 3]
```diff
@@ -338,6 +353,107 @@ def load(file_path):
...
+        # calib data should be set if the forecaster is just loaded
+        assert calib_data is not None, "You must set a `calib_data` for quantization."
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:408]: add a `if` logic here. `xxx and approach=="static`

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py (hunk 1 above)
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/autots/tspipeline.py:381]: 4. numpy ndarray, ...

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 4]
```diff
@@ -338,6 +353,107 @@ def load(file_path):
...
+                                                  torch.from_numpy(calib_data[1])))
+        if isinstance(calib_data, types.FunctionType):
+            calib_data = calib_data(self._best_config)
+
+        if isinstance(val_data, tuple):
+            val_data = DataLoader(TensorDataset(torch.from_numpy(val_data[0]),
+                                                torch.from_numpy(val_data[1])))
+        if isinstance(val_data, types.FunctionType):
+            val_data = val_data(self._best_config)
+
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:421]: add `TSDataset` support here

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 5]
```diff
@@ -338,6 +353,107 @@ def load(file_path):
...
+        accuracy_criterion = None
+        if relative_drop and absolute_drop:
+            raise ValueError("Please unset either `relative_drop` or `absolute_drop`.")
+        if relative_drop:
+            accuracy_criterion = {'relative': relative_drop, 'higher_is_better': False}
+        if absolute_drop:
+            accuracy_criterion = {'absolute': absolute_drop, 'higher_is_better': False}
+
+        from bigdl.nano.pytorch.trainer import Trainer
+        num_processes = max(1, torch.get_num_threads()//8)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:437]: no need for `num_process` here

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 6]
```diff
@@ -338,6 +353,122 @@ def load(file_path):
...
+               decide when to exit. "timeout=0, max_trials=1" means it will try quantization
+               only once and return satisfying best model.
+        """
+        from torch.utils.data import DataLoader, TensorDataset
+        from bigdl.chronos.data import TSDataset
+        # check model support for quantization
+        from bigdl.chronos.autots.utils import check_quantize_available
+        check_quantize_available(self._best_model.model)
+        # calib data should be set if the forecaster is just loaded
+        assert calib_data is not None and approach.startswith("static"), \
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:416]: Use `if xxxx` `raise` logic. Make sure "dynamic" can be used here

File: python/chronos/test/bigdl/chronos/autots/test_tspipeline.py [hunk 7]
```diff
@@ -174,5 +174,77 @@ def test_tsppl_mixed_data_type_usage(self):
...
+        with pytest.raises(AssertionError):
+            tsppl_tcn.quantize(calib_data=(calib_x, calib_y),
+                               metric='smape',
+                               approach='dynamic')
+
+        from bigdl.chronos.data.repo_dataset import get_public_dataset
+        train_tsdata, _, test_tsdata = get_public_dataset('network_traffic', val_ratio=0)
+        train_tsdata.roll(lookback=10, horizon=2)
+        test_tsdata.roll(lookback=10, horizon=2)
+        tsppl_tcn._best_config.update({'selected_features': []})
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/autots/test_tspipeline.py:241]: add a comment to illustrate this hacking change

Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/autots/test_tspipeline.py:241]: add a comment to illustrate this hacking change

**Comment by TheaperDeng** [comment]
generate a document just for validation checking

**Comment by liangs6212** [comment]
https://smalldl.readthedocs.io/en/tspipeline-quantization/doc/PythonAPI/Chronos/autotsestimator.html#tspipeline

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 8]
```diff
@@ -338,6 +353,126 @@ def load(file_path):
...
+                 conf=None,
+                 framework='pytorch_fx',
+                 approach='static',
+                 tuning_strategy='bayesian',
+                 relative_drop=None,
+                 absolute_drop=None,
+                 timeout=0,
+                 max_trials=1):
+        """
+        Quantization TSPipeline.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:369]: give a blank line here might solve the formatting problem in document

File: python/chronos/src/bigdl/chronos/autots/tspipeline.py [hunk 9]
```diff
@@ -338,6 +353,126 @@ def load(file_path):
...
+        from bigdl.chronos.autots.utils import check_quantize_available
+        check_quantize_available(self._best_model.model)
+        # calib data should be set if the forecaster is just loaded
+        if calib_data is None and approach.startswith("static"):
+            raise RuntimeError("You must set a `calib_data` "
+                               "for quantization When you use 'static'.")
+        elif calib_data and approach.startswith("dynamic"):
+            raise RuntimeError("`calib_data` should be None When you use 'dynamic'.")
+
+        # change data tuple to dataloader
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/autots/tspipeline.py:422]: line 423 - 431 and line 433 - 441 has nearly the same pattern, maybe we can use a function to clean the code.
```python
calib_data = preprocess_quantize_data(calib_data)
val_data = preprocess_quantize_data(val_data)
```

**Review by TheaperDeng — APPROVED** [review]
LGTM

**Comment by liangs6212** [comment]
http://10.112.231.51:18889/view/BigDL-PR-Validation/job/BigDL-Chronos-PR-Validation/318/

**Pull request merged**